pull request iss-refactor-checkpointing to feature/iss

federicamontes opened the pull request:
- file(configuration.c, configuration.h) : refactored code, changed parameters name for compliance reasons from forced_full_ckpt_period to ckpt_forced_full_period 

- file(checkpoints.c) : removed method is_next_ckpt_incremental() and moved it to incremental_state_saving.c
* the method switches from PERIODIC_STATE_SAVING to INCREMENTAL_STATE_SAVING  to take a partial log and after ckpt_period events it switches back to PERIODIC_STATE_SAVING to take a full log

- file(recoverable.c) : removed comments and removed logged_state->dirty_bitmap_size from log size computation in case of incremental state saving

- moved incremental_state_saving.h from include/ to mm/

- file(state.c): the method LogState switches also on INCREMENTAL_STATE_SAVING which is set in is_next_ckpt_incremental() (file: incremental_state_saving.c)

Commit:
Merge branch 'feature/iss' into iss-refactor-checkpointing

Commit:
refactor(configuration.c): added forced_full_ckpt_period option to support the transition from incremental to full log

Commit:
refactor(configuration.h) added forced_full_ckpt_period and incremental_state_saving as log mode

Commit:
refactor(checkpoints.c): moved some parts of full log and full restore in separate function

* chunks_not_used and scan_recoverable_area return bool values to decide whether to skip the for loop respectively in log_full and restore_full

* complete_full_log and complete_full_restore finish the full log logging the chunks inside the areas

* is_next_ckpt_incremental decides whether we take incremental snapshots with every ckpt_period or if enough incremental logs have been taken so to force a full log

Commit:
refactor(incremental_state_saving.c): included a header file and commented two structs

Commit:
refactor(recoverable.c): changed incremental case log size

Commit:
refactor(state.c): added INCREMENTAL_STATE_SAVING case

Commit:
feature(incremental_state_saving.h) added new header

* added definition of __per_lp_iss_metadata and __partition_tree_node

* added is_next_ckpt_incremental signature

Commit:
refactor(configuration.c): changed config parameter from forced_full_ckpt_period to ckpt_forced_full_period for compliance reasons

Commit:
refactor(configuration.h): changed parameter forced_full_ckpt_period to ckpt_forced_full_period for compliance reasons

Commit:
refactor(incremental_state_saving.h) removed header file incremental_state_saving.h from include/ directory

Commit:
refactor(checkpoints.c): removed method is_next_ckpt_incremental

Commit:
refactor(incremental_state_saving.c): added method is_next_ckpt_incremental

* the method switches from PERIODIC_STATE_SAVING to INCREMENTAL_STATE_SAVING  to take a partial log and after ckpt_period events it switches back to PERIODIC_STATE_SAVING to take a full log

Commit:
refactor(recoverable.c) removed comments and logged_state->dirty_bitmap_size from the computation of log size for the incremental case

Commit:
refactor(incremental_state_saving.h): moved header file incremental_state_saving.h in mm/ directory

RomoloMarotta commented on src/core/configuration.c:
```diff
@@ -47,6 +48,7 @@ static struct argp_option options[] = {
   {"ckpt-period",         CKPT_PERIOD_KEY          , "#EVENTS" ,  0                  ,  "Number of events to be forward-executed before taking a full-snapshot"   , 0 },
   {"ckpt-fossil-period",  CKPT_FOSSIL_PERIOD_KEY   , "#EVENTS" ,  0                  ,  "Number of events to be executed before collection committed snapshot"   , 0 },
   {"ckpt-autonomic-period",  CKPT_AUTONOMIC_PERIOD_KEY, 0         ,  OPTION_ARG_OPTIONAL,  "Enable autonomic checkpointing period"   , 0 },
+  {"ckpt_forced_full_period",  CKPT_FORCED_FULL_PERIOD_KEY, 0         ,  OPTION_ARG_OPTIONAL,  "Number of incremental checkpoints before taking a full log"   , 0 },
```
we need a value for ckpt_forced_full_period

```suggestion
  {"ckpt_forced_full_period",  CKPT_FORCED_FULL_PERIOD_KEY, "#EVENTS"         ,  0,  "Number of incremental checkpoints before taking a full log"   , 0 },
```

RomoloMarotta commented on src/mm/state.c:
```diff
@@ -150,6 +150,8 @@ bool LogState(unsigned int lid) {
 				LPS[lid]->from_last_ckpt = 0;
 			}
 			break;
+		case INCREMENTAL_STATE_SAVING:
+			take_snapshot = true;
 
 		default:
```
This has missing break, so it executes rootsim_error.
It should have the same case as PERIODIc_STATE_SAVING.

see the following

```suggestion

		case INCREMENTAL_STATE_SAVING:
		case PERIODIC_STATE_SAVING:
			if(LPS[lid]->from_last_ckpt >= LPS[lid]->ckpt_period) {
				take_snapshot = true;
				LPS[lid]->from_last_ckpt = 0;
			}
			break;
		default:
```

RomoloMarotta commented on src/mm/incremental_state_saving.c:
```diff
@@ -33,11 +34,29 @@ typedef struct __partition_tree_node{
+
+bool is_next_ckpt_incremental() {
+
+	if (pdes_config.checkpointing == PERIODIC_STATE_SAVING) {
+		pdes_config.checkpointing == INCREMENTAL_STATE_SAVING; /// change ckpt mode to incremental state saving
+		pdes_config.ckpt_forced_full_period = pdes_config.ckpt_period; /// set iss period
+	}
+
+	if (iss_states[current_lp].iss_counter == pdes_config.ckpt_forced_full_period) { /// after iss_counter incremental state saving take a full log
+		pdes_config.checkpointing == PERIODIC_STATE_SAVING; /// change ckpt mode to periodic state saving
+		return false;
+	}
```
this looks simpler

```suggestion
	if (pdes_config.checkpointing == PERIODIC_STATE_SAVING)  
		return false;

	if (iss_states[current_lp].iss_counter == pdes_config.ckpt_forced_full_period) { /// after iss_counter incremental state saving take a full log
		iss_states[current_lp].iss_counter++; /// its time to take a full snapshot
		return false;
	}
```

federicamontes replied:
```suggestion
	if (pdes_config.checkpointing == PERIODIC_STATE_SAVING)  
		return false;
	if (iss_states[current_lp].iss_counter == pdes_config.ckpt_forced_full_period) { /// after iss_counter incremental state saving take a full log
		iss_states[current_lp].iss_counter = 0; /// its time to take a full snapshot
		return false;
	}
```

RomoloMarotta commented on src/mm/checkpoints.c:
```diff
@@ -38,6 +38,81 @@
+
+
+
+bool chunks_not_used(malloc_area **m_area, int *bitmap_blocks) {
```
* why a pointer of pointer?
* better name
```suggestion
bool check_not_used_chunk_and_clean(malloc_area *m_area, int *bitmap_blocks) {
```

RomoloMarotta commented on src/mm/checkpoints.c:
```diff
@@ -38,6 +38,81 @@
+		if ((*m_area)->use_bitmap != NULL) {
+			for(j = 0; j < *bitmap_blocks; j++)
+				(*m_area)->dirty_bitmap[j] = 0;
+		}
```
```suggestion
	*bitmap_blocks = m_area->num_chunks / NUM_CHUNKS_PER_BLOCK;
	if (*bitmap_blocks < 1)
		*bitmap_blocks = 1;

	// Check if there is at least one chunk used in the area
	if(m_area->alloc_chunks == 0){

		m_area->dirty_chunks = 0;
		m_area->state_changed = 0;

		if (m_area->use_bitmap != NULL) {
			for(j = 0; j < *bitmap_blocks; j++)
				m_area->dirty_bitmap[j] = 0;
		}
```

RomoloMarotta commented on src/mm/checkpoints.c:
```diff
@@ -107,24 +183,7 @@ void *log_full(int lid) {
-
-			continue;
-		}
+		if (chunks_not_used(&m_area, &bitmap_blocks)) continue;
```
```suggestion
		if (chunks_not_used(m_area, &bitmap_blocks)) continue;
```

RomoloMarotta commented on src/mm/checkpoints.c:
```diff
@@ -38,6 +38,81 @@
+	int j, k, idx;
+
+	chunk_size = (*m_area)->chunk_size;
+	RESET_BIT_AT(chunk_size, 0);	// ckpt Mode bit
+	RESET_BIT_AT(chunk_size, 1);	// Lock bit
+
+	// Check whether the area should be completely copied (not on a per-chunk basis)
+	// using a threshold-based heuristic
+	if(CHECK_LOG_MODE_BIT(*m_area)){
+
+		// If the malloc_area is almost (over a threshold) full, copy it entirely
+		memcpy(*ptr, (*m_area)->area, (*m_area)->num_chunks * chunk_size);
+		*ptr = (void*)((char*) *ptr + (*m_area)->num_chunks * chunk_size);
+
+	} else {
+		// Copy only the allocated chunks
+		for(j = 0; j < bitmap_blocks; j++){
+
+			// Check the allocation bitmap on a per-block basis, to enhance scan speed
+			if((*m_area)->use_bitmap[j] == 0) {
+				// Empty (no-chunks-allocated) block: skip to the next
+				continue;
+
+			} else {
+				// At least one chunk is allocated: per-bit scan of the block is required
+				for(k = 0; k < NUM_CHUNKS_PER_BLOCK; k++){
+
+					if(CHECK_BIT_AT((*m_area)->use_bitmap[j], k)){
+
+						idx = j * NUM_CHUNKS_PER_BLOCK + k;
+						memcpy(*ptr, (void*)((char*)(*m_area)->area + (idx * chunk_size)), chunk_size);
+						*ptr = (void*)((char*) *ptr + chunk_size);
+
+					}
+				}
+			}
+		}
+
+	}
+}
```
why pointer of pointer for m_area?

```suggestion
void log_all_marea_chunks(malloc_area *m_area, void **ptr, int bitmap_blocks) {

	size_t chunk_size;
	int j, k, idx;

	chunk_size = m_area->chunk_size;
	RESET_BIT_AT(chunk_size, 0);	// ckpt Mode bit
	RESET_BIT_AT(chunk_size, 1);	// Lock bit

	// Check whether the area should be completely copied (not on a per-chunk basis)
	// using a threshold-based heuristic
	if(CHECK_LOG_MODE_BIT(m_area)){

		// If the malloc_area is almost (over a threshold) full, copy it entirely
		memcpy(*ptr, m_area->area, m_area->num_chunks * chunk_size);
		*ptr = (void*)((char*) *ptr + m_area->num_chunks * chunk_size);

	} else {
		// Copy only the allocated chunks
		for(j = 0; j < bitmap_blocks; j++){

			// Check the allocation bitmap on a per-block basis, to enhance scan speed
			if(m_area->use_bitmap[j] == 0) {
				// Empty (no-chunks-allocated) block: skip to the next
				continue;

			} else {
				// At least one chunk is allocated: per-bit scan of the block is required
				for(k = 0; k < NUM_CHUNKS_PER_BLOCK; k++){

					if(CHECK_BIT_AT(m_area->use_bitmap[j], k)){

						idx = j * NUM_CHUNKS_PER_BLOCK + k;
						memcpy(*ptr, (void*)((char*)m_area->area + (idx * chunk_size)), chunk_size);
						*ptr = (void*)((char*) *ptr + chunk_size);

					}
				}
			}
		}

	}
}
```

RomoloMarotta commented on src/mm/checkpoints.c:
```diff
@@ -133,42 +192,16 @@ void *log_full(int lid) {
-				} else {
-					// At least one chunk is allocated: per-bit scan of the block is required
-					for(k = 0; k < NUM_CHUNKS_PER_BLOCK; k++){
+			complete_full_log(&m_area, &ptr, bitmap_blocks);
```
```suggestion
			log_all_marea_chunks(m_area, &ptr, bitmap_blocks);
```

RomoloMarotta commented on src/mm/checkpoints.c:
```diff
@@ -273,35 +388,12 @@ void restore_full(int lid, void *ckpt) {
-					m_area->dirty_bitmap[j] = 0;
-			}
-			m_area->last_access = recoverable_state[lid]->timestamp;
+		if (scan_recoverable_area(&m_area, &ptr, &bitmap_blocks, &restored_areas, &recoverable_state[lid])) continue;
```
```suggestion
		if (check_marea_rebuild(m_area, &ptr, &bitmap_blocks, &restored_areas, &recoverable_state[lid])) continue;
```

RomoloMarotta commented on src/mm/checkpoints.c:
```diff
@@ -219,6 +252,90 @@ void *log_state(int lid) {
 
 
 
+bool scan_recoverable_area(malloc_area **m_area, void **ptr, int *bitmap_blocks, int *restored_areas, malloc_state **recoverable_state) {
```
why pointer of pointer?
```suggestion
bool check_marea_rebuild(malloc_area *m_area, void **ptr, int *bitmap_blocks, int *restored_areas, malloc_state **recoverable_state) {
```

RomoloMarotta commented on src/mm/checkpoints.c:
```diff
@@ -319,46 +411,15 @@ void restore_full(int lid, void *ckpt) {
-				if(bitmap == 0){
-					// Empty (no-chunks-allocated) block: skip to the next
-					continue;
+			complete_full_restore(&m_area, &ptr, bitmap_blocks);
```
```suggestion
			restore_marea_chunk(m_area, &ptr, bitmap_blocks);
```

RomoloMarotta commented on src/mm/checkpoints.c:
```diff
@@ -219,6 +252,90 @@ void *log_state(int lid) {
+}
+
+
+void complete_full_restore(malloc_area **m_area, void **ptr, int bitmap_blocks) {
```
```suggestion
void restore_marea_chunk(malloc_area *m_area, void **ptr, int bitmap_blocks) {
```

federicamontes committed:
Update src/core/configuration.c

added arg #EVENTS for ckpt_forced_full_period

Co-authored-by: RomoloMarotta <romolo.marotta@gmail.com>

federicamontes committed:
Update src/mm/state.c

collapsed INCREMENTAL_STATE_SAVING in PERIODIC_STATE_SAVING

Co-authored-by: RomoloMarotta <romolo.marotta@gmail.com>

federicamontes committed:
Update src/mm/checkpoints.c

changed name of method chunks_not_used to check_not_used_chunks_and_clean and signature, first parameter is malloc_area * and not malloc_area **

Co-authored-by: RomoloMarotta <romolo.marotta@gmail.com>

federicamontes committed:
Update src/mm/checkpoints.c

changed m_area to malloc_area * instead of malloc_area **

Co-authored-by: RomoloMarotta <romolo.marotta@gmail.com>

federicamontes committed:
Update src/mm/checkpoints.c

corrected parameter usage, pass malloc_area * and not the address of malloc_area

Co-authored-by: RomoloMarotta <romolo.marotta@gmail.com>

federicamontes committed:
Update src/mm/checkpoints.c

changed method name from complete_full_log to log_all_marea_chunks

Co-authored-by: RomoloMarotta <romolo.marotta@gmail.com>

federicamontes committed:
Update src/mm/checkpoints.c

corrected usage of malloc_area from malloc_area ** to malloc_area *

Co-authored-by: RomoloMarotta <romolo.marotta@gmail.com>

federicamontes committed:
Update src/mm/checkpoints.c

changed method name from scan_recoverable_area to check_marea_rebuild

Co-authored-by: RomoloMarotta <romolo.marotta@gmail.com>

federicamontes committed:
Update src/mm/checkpoints.c

changed parameter usage from malloc_area ** to malloc_area *

Co-authored-by: RomoloMarotta <romolo.marotta@gmail.com>

federicamontes committed:
Update src/mm/checkpoints.c

changed method name from complete_full_restore to restore_marea_chunk

Co-authored-by: RomoloMarotta <romolo.marotta@gmail.com>

federicamontes committed:
Update src/mm/checkpoints.c

corrected parameter usage from malloc_area ** to malloc_area *

Co-authored-by: RomoloMarotta <romolo.marotta@gmail.com>

federicamontes committed:
Update src/mm/incremental_state_saving.c

edited iss_counter increment to iss_counter reset because in that if body we force a full log

RomoloMarotta commented on src/core/configuration.c:
```diff
@@ -47,6 +48,7 @@ static struct argp_option options[] = {
   {"ckpt-period",         CKPT_PERIOD_KEY          , "#EVENTS" ,  0                  ,  "Number of events to be forward-executed before taking a full-snapshot"   , 0 },
   {"ckpt-fossil-period",  CKPT_FOSSIL_PERIOD_KEY   , "#EVENTS" ,  0                  ,  "Number of events to be executed before collection committed snapshot"   , 0 },
   {"ckpt-autonomic-period",  CKPT_AUTONOMIC_PERIOD_KEY, 0         ,  OPTION_ARG_OPTIONAL,  "Enable autonomic checkpointing period"   , 0 },
+  {"ckpt_forced_full_period",  CKPT_FORCED_FULL_PERIOD_KEY, "#EVENTS"         ,  0,  "Number of incremental checkpoints before taking a full log"   , 0 },
```
```suggestion
  {"ckpt_forced_full_period",  CKPT_FORCED_FULL_PERIOD_KEY, "#CHECKPOINTS"         ,  0,  "Number of incremental checkpoints before taking a full log"   , 0 },
```

RomoloMarotta commented on src/mm/incremental_state_saving.c:
```diff
@@ -33,11 +34,26 @@ typedef struct __partition_tree_node{
+
+	if (pdes_config.checkpointing == PERIODIC_STATE_SAVING)  
+		return false;
+	if (iss_states[current_lp].iss_counter == pdes_config.ckpt_forced_full_period) { /// after iss_counter incremental state saving take a full log
```
```suggestion
	if (iss_states[current_lp].iss_counter++ == pdes_config.ckpt_forced_full_period) { /// after iss_counter incremental state saving take a full log
```

RomoloMarotta commented on src/mm/state.c:
```diff
@@ -144,13 +144,17 @@ bool LogState(unsigned int lid) {
 			take_snapshot = true;
 			break;
 
+		case PERIODIC_STATE_SAVING:
+			if(LPS[lid]->from_last_ckpt >= LPS[lid]->ckpt_period) {
+				take_snapshot = true;
```
```suggestion
```

RomoloMarotta commented on src/mm/checkpoints.c:
```diff
@@ -219,6 +252,90 @@ void *log_state(int lid) {
+
+	int j;
+
+	*bitmap_blocks = (*m_area)->num_chunks / NUM_CHUNKS_PER_BLOCK;
+	if(*bitmap_blocks < 1)
+		*bitmap_blocks = 1;
+
+	if(*restored_areas == (*recoverable_state)->busy_areas || (*m_area)->idx != ((malloc_area*) *ptr)->idx){
+
+		(*m_area)->dirty_chunks = 0;
+		(*m_area)->state_changed = 0;
+		if ((*m_area)->use_bitmap != NULL){
+			for(j = 0; j < *bitmap_blocks; j++) {
+			  (*m_area)->dirty_bitmap[j] = 0;
+			}
+		}
+		(*m_area)->alloc_chunks = 0;
+		(*m_area)->next_chunk = 0;
+		RESET_LOG_MODE_BIT(*m_area);
+		RESET_AREA_LOCK_BIT(*m_area);
+		if ((*m_area)->use_bitmap != NULL) {
+			for(j = 0; j < *bitmap_blocks; j++)
+				(*m_area)->use_bitmap[j] = 0;
+			for(j = 0; j < *bitmap_blocks; j++)
+				(*m_area)->dirty_bitmap[j] = 0;
+		}
+		(*m_area)->last_access = (*recoverable_state)->timestamp;
+
+		return true;
+	}
+
+	return false;
```
There is (*m_area).
Probably this does not compile

federicamontes replied:
```suggestion
	*bitmap_blocks = m_area->num_chunks / NUM_CHUNKS_PER_BLOCK;
	if(*bitmap_blocks < 1)
		*bitmap_blocks = 1;

	if(*restored_areas == (*recoverable_state)->busy_areas || m_area->idx != ((malloc_area*) *ptr)->idx){

		m_area->dirty_chunks = 0;
		m_area->state_changed = 0;
		if (m_area->use_bitmap != NULL){
			for(j = 0; j < *bitmap_blocks; j++) {
			  m_area->dirty_bitmap[j] = 0;
			}
		}
		m_area->alloc_chunks = 0;
		m_area->next_chunk = 0;
		RESET_LOG_MODE_BIT(m_area);
		RESET_AREA_LOCK_BIT(m_area);
		if (m_area->use_bitmap != NULL) {
			for(j = 0; j < *bitmap_blocks; j++)
				m_area->use_bitmap[j] = 0;
			for(j = 0; j < *bitmap_blocks; j++)
				m_area->dirty_bitmap[j] = 0;
		}
		m_area->last_access = (*recoverable_state)->timestamp;

		return true;
	}

	return false;
```

RomoloMarotta commented on src/mm/checkpoints.c:
```diff
@@ -219,6 +252,90 @@ void *log_state(int lid) {
+	unsigned int bitmap;
+
+	chunk_size = (*m_area)->chunk_size;
+	RESET_BIT_AT(chunk_size, 0);
+	RESET_BIT_AT(chunk_size, 1);
+
+	// Check how the area has been logged
+	if(CHECK_LOG_MODE_BIT(*m_area)){
+		// The area has been entirely logged
+		memcpy((*m_area)->area, ptr, (*m_area)->num_chunks * chunk_size);
+		*ptr = (void*)((char*) *ptr + (*m_area)->num_chunks * chunk_size);
+
+	} else {
+		// The area was partially logged.
+		// Logged chunks are the ones associated with a used bit whose value is 1
+		// Their number is in the alloc_chunks counter
+		for(j = 0; j < bitmap_blocks; j++){
+
+			bitmap = (*m_area)->use_bitmap[j];
+
+			// Check the allocation bitmap on a per-block basis, to enhance scan speed
+			if(bitmap == 0){
+				// Empty (no-chunks-allocated) block: skip to the next
+				continue;
+
+			} else {
+				// Scan the bitmap on a per-bit basis
+				for(k = 0; k < NUM_CHUNKS_PER_BLOCK; k++){
+
+					if(CHECK_BIT_AT(bitmap, k)){
+
+						idx = j * NUM_CHUNKS_PER_BLOCK + k;
+						memcpy((void*)((char*)(*m_area)->area + (idx * chunk_size)), *ptr, chunk_size);
+						*ptr = (void*)((char*) *ptr + chunk_size);
+
+					}
+				}
+			}
+		}
+	}
```
(*m_area)not needed

federicamontes replied:
```suggestion
	chunk_size = m_area->chunk_size;
	RESET_BIT_AT(chunk_size, 0);
	RESET_BIT_AT(chunk_size, 1);

	// Check how the area has been logged
	if(CHECK_LOG_MODE_BIT(m_area)){
		// The area has been entirely logged
		memcpy(m_area->area, ptr, m_area->num_chunks * chunk_size);
		*ptr = (void*)((char*) *ptr + m_area->num_chunks * chunk_size);

	} else {
		// The area was partially logged.
		// Logged chunks are the ones associated with a used bit whose value is 1
		// Their number is in the alloc_chunks counter
		for(j = 0; j < bitmap_blocks; j++){

			bitmap = m_area->use_bitmap[j];

			// Check the allocation bitmap on a per-block basis, to enhance scan speed
			if(bitmap == 0){
				// Empty (no-chunks-allocated) block: skip to the next
				continue;

			} else {
				// Scan the bitmap on a per-bit basis
				for(k = 0; k < NUM_CHUNKS_PER_BLOCK; k++){

					if(CHECK_BIT_AT(bitmap, k)){

						idx = j * NUM_CHUNKS_PER_BLOCK + k;
						memcpy((void*)((char*)m_area->area + (idx * chunk_size)), *ptr, chunk_size);
						*ptr = (void*)((char*) *ptr + chunk_size);

					}
				}
			}
		}
	}
```

federicamontes reviewed (commented):
approved changes in configuration.c and state.c
edited malloc_area ** issue in checkpoints.c

RomoloMarotta committed:
Update src/core/configuration.c

refactor(configuration.c): renamed a field in the command line description

RomoloMarotta committed:
Update src/mm/checkpoints.c

bugfix(checkpoints): fixed compilation issue

Co-authored-by: Federica Montesano <federica.montesano42@gmail.com>

RomoloMarotta committed:
Update src/mm/checkpoints.c

bugfix(checkpoints): fixed compilation issue

Co-authored-by: Federica Montesano <federica.montesano42@gmail.com>

RomoloMarotta committed:
Update src/mm/incremental_state_saving.c

bugfix(iss): fixed missing increments for iss counter

RomoloMarotta committed:
Update src/mm/state.c

bugfix(state): fixed compilation issues

RomoloMarotta committed:
bugfix(checkpoints): fixed wrong name of a function

Signed-off-by: Romolo Marotta <romolo.marotta@gmail.com>
